### DISCUSSION TITLE
Checking http implementation

### DISCUSSION THREAD

**Discussion by trikko** [the-benchmarker/web-frameworks#7422] [Ideas]
It seems that some frameworks are just build to win the benchmark but they are not really usable. 

I think we should test for example the frameworks against invalid or malformed requests in many ways.

But even valid requests seem not correctly handled by many frameworks.

For example: is the "expect" header handled? If you post a file with curl, it sends an "expect" header and wait for server to send a 100-continue reply before sending the file body.

Or also: what if I send many requests together without waiting the reply from server ("http pipeline") or if I send a request in separated fragments like "GE" (sleep) "T / HTT" (sleep) "P/1.1" etc.

What if client post more data than declared in "content-lenght"?

**Comment by waghanza** [comment]
It will be very helpful if you list what is expected, and seems weird in some frameworks.

I can add some tests and start some modification on some implementations

**Comment by trikko** [comment]
There are numerous tests we can perform. I would start with some basic checks to ensure that we have a functioning server.

For instance, requests that begin with lines like the following should return a "400 Bad Request" response:
```
FAKE / HTTP/1.1
GETTT / HTTP/1.1
GET / HTTP/1.3
GET / HTTP/5.1
BLAH
```

Every HTTP/1.1 request must include a "Host" header; otherwise, it is considered a bad request.

The following request instead should return a `400 Bad Request` or a HTTP/1.0 response: 
```
GET / HTTP/1.0
```

I would also verify if some basic headers are parsed correctly. For example:
- `Expect: 100-continue`: If an HTTP/1.1 client receives this with a POST method, it should reply with a `100 Continue` or `417 Expectation Failed`.
- Both `Connection: keep-alive` and `Connection: close` must be supported.
- Without a `Connection` header, it defaults to `keep-alive` for HTTP/1.1 and `close` for HTTP/1.0, and the server must respond accordingly.
- What happens if I send `Connection: blah`?
- What happens if I send a header without a value like `X-header:`?
- Header fields are case-insensitive. So, both `connection: close` and `CONnecTION: close` should work.
- `Content-Length` must be parsed to read the exact amount of data from the client. Any extra data belongs to the next request in the pipeline (if the connection is `keep-alive`. If the connection is `close`, it's a bad request).
- What happens if I send `Content-Length: -5`?

I would also test pipelining, which involves sending multiple requests at once (with HTTP/1.1 and `Connection: keep-alive`) without waiting for the server's reply. The server must reply in the same order as the requests were received.

There are also other desirable behaviors to test. For example, what happens if a client opens a connection and doesn’t send any data? After a while, I think the connection should be closed to avoid occupying resources. The same goes for keep-alive: there should be a timeout after which the connection is interrupted if no more requests are received.

And so on...

**Comment by cyrusmsk** [comment]
Found this tool on GH. Could be useful for testing probably https://github.com/Orange-OpenSource/hurl

**Comment by waghanza** [comment]
@MDA2AV I've checked you project `Http11Probe` on `actix`, and I have
```
Score: 83/135 (52 failed, 8 warnings)  49 errors  54 unscored  (215 tests, 35,4s)
```
but I can see 
<img width="813" height="45" alt="image" src="https://github.com/user-attachments/assets/657a9000-030a-4501-8162-635cc2fba9fa" />
on https://www.http-probe.com/probe-results/

The command used is `dotnet run --project src/Http11Probe.Cli -- --host 172.17.0.2 --port 3000` from **master**

**Reply by MDA2AV** [reply]
just tested it locally
Score: 139/161 (22 failed, 23 warnings)  54 unscored  (215 tests, 78.2s)

did you run it against the actix implementation on the http11probe platform or your actix implementation?

**Reply by waghanza** [reply]
ah, you mean spec compliance must be run on a specific impl ? 

I mean I want to evaluate compliance of actix, and the impl is https://github.com/the-benchmarker/web-frameworks/blob/develop/rust/actix/src/main.rs, but maybe we should make some update

The global idea is to test compliance of each framework in here

**Reply by MDA2AV** [reply]
Ah, well it works partially, some tests are for CONNECT, POST, HEAD and those you don't implement in your benchmark so naturally the results won't match what you see on Http11Probe so you need to evaluate only the tests that make sense. 

For example some tests(normalization) expect that the target server replies an echo of all the headers received to check how the framework parses headers, those will naturally fail if you dont implement that. There are tests for websockets, caching, cookies etc and those may not be enabled by default.

https://github.com/MDA2AV/Http11Probe/blob/main/src/Servers/ActixServer/src/main.rs
this is the Http11Probe implementation for actix, has some extra logic

In your case I'd look only for mostly compliance, this section - https://www.http-probe.com/compliance/

**Reply by cyrusmsk** [reply]
@waghanza both repos are full AI-slop
I would use any results from both with a huge grant of salt

**Reply by waghanza** [reply]
> @waghanza both repos are full AI-slop
> I would use any results from both with a huge grant of salt

Well not sure going to publish any results @cyrusmsk , but willl take a look (even if AI vibe codes)

**Reply by MDA2AV** [reply]
> > @waghanza both repos are full AI-slop
> > I would use any results from both with a huge grant of salt
> 
> Well not sure going to publish any results @cyrusmsk , but willl take a look (even if AI vibe codes)

Lol, don't worry this is not a vibe coded project, plus you can find detailed results for each tests (request and response) and chain of thought reasoning for each test backed by official RFC docs

**Reply by waghanza** [reply]
I am not sure we should yell if AI is used or not (vibe coded or AI optimized) but maybe it could appear like a research project as for now.

It could lead to curiosty, but maybe not mature enough to use at large scall.

AI is a tool (so it could be usefull or useless depending on how we use), the idea is to stick to science to make sure to add clarity on digital world, at least this is the pledge here @cyrusmsk @MDA2AV 

War is not helpful 😜
